### PR TITLE
Ghosts can make candles flicker via boo or mouseclick.

### LINKED ISSUE #34999: Allow Ghosts to Make Candles Flicker
#### Description of issue
I just think it would be neat if ghosts could click a candle to make it flicker in the same way they're able to do with lights. It would also provide a fun little benign way for ghosts to interact with the chaplain since he gets plenty of candles.

### PR DESCRIPTION
**Pull request by Eneocho**
## What this does
Ghosts can haunt candles just like they can haunt wall lights.
Closes #34999.
[tested] and it didn't break
Code shamelessly stolen and modified from lighting.dm
:cl:
 * rscadd: Ghosts can make candles flicker via mouseclick or Boo!

### CODE REVIEW

**Comment by pedr69** [comment]
but how does it re-ignite?

**Comment by Eneocho** [comment]
ghostly welders

**Comment by Hinaichigo** [comment]
Maybe they could "almost sputter out" (have the light level drop to something very low) rather than turning off entirely.

**Comment by Eneocho** [comment]
I considered it but that's a rework to candle lighting and icons that I didn't really feel was necessary. This implementation is simple and works. Plus the candles relighting feels more supernatural.

File: code/game/objects/items/candle.dm [hunk 1]
```diff
@@ -69,6 +70,32 @@
...
+
+/obj/item/candle/spook(mob/dead/observer/ghost)
+	if(..(ghost, TRUE))
+		flicker()
+
+/obj/item/candle/attack_ghost(mob/user)
+	if(!can_spook())
+		return
+	src.add_hiddenprint(user)
+	src.flicker(1)
```
Comment (Kurfursten) [inline code/game/objects/items/candle.dm]: no need for src

File: code/game/objects/items/candle.dm [hunk 2]
```diff
@@ -69,6 +70,32 @@
...
+	investigation_log(I_GHOST, "|| was made to flicker by [key_name(user)][user.locked_to ? ", who was haunting [user.locked_to]" : ""]")
+	return
```
Comment (Kurfursten) [inline code/game/objects/items/candle.dm]: no trailing returns

File: code/game/objects/items/candle.dm [hunk 3]
```diff
@@ -69,6 +70,32 @@
...
+	if(flickering)
+		return
+	flickering = 1
+	spawn(0)
+		if(lit)
+			for(var/i = 0; i < amount; i++)
+				lit = 0
+				update_icon()
+				sleep(rand(3,9))
+				lit = 1
```
Comment (Kurfursten) [inline code/game/objects/items/candle.dm]: you could get a better flicker effect by instead adjusting the light power or radius (or both)

File: code/game/objects/items/candle.dm [hunk 4]
```diff
@@ -69,6 +70,32 @@
 		processing_objects.Add(src)
 		update_icon()
 
+/obj/item/candle/proc/flicker(var/amount = rand(5, 15))
+	if(flickering)
+		return
+	flickering = 1
+	spawn(0)
```
Comment (Kurfursten) [inline code/game/objects/items/candle.dm]: is this spawn necessary...? the function ends

File: code/game/objects/items/candle.dm (hunk 1 above)
Comment (Eneocho) [reply code/game/objects/items/candle.dm]: you sure? I copypasted this (and removed the wall light-specific parts) from lighting.dm, that's how it's implemented there

File: code/game/objects/items/candle.dm (hunk 4 above)
Comment (Eneocho) [reply code/game/objects/items/candle.dm]: I assume so, that's how it was implemented on the lighting.dm file.

File: code/game/objects/items/candle.dm (hunk 3 above)
Comment (Eneocho) [reply code/game/objects/items/candle.dm]: Fine, I'll see what I can do.

File: code/game/objects/items/candle.dm (hunk 2 above)
Comment (Eneocho) [reply code/game/objects/items/candle.dm]: Will remove once I test some shit.

File: code/game/objects/items/candle.dm (hunk 1 above)
Comment (Kurfursten) [reply code/game/objects/items/candle.dm]: certain, src just refers to self

File: code/game/objects/items/candle.dm (hunk 4 above)
Comment (Kurfursten) [reply code/game/objects/items/candle.dm]: it doesn't appear to do anything at all here

Comment (Eneocho) [reply code/game/objects/items/candle.dm]: tested it, the proc doesn't compile if the spawn ain't there

Comment (Hinaichigo) [reply code/game/objects/items/candle.dm]: What's the error message?

File: code/game/objects/items/candle.dm (hunk 3 above)
Comment (Eneocho) [reply code/game/objects/items/candle.dm]: done

File: code/game/objects/items/candle.dm (hunk 4 above)
Comment (Eneocho) [reply code/game/objects/items/candle.dm]: `code\game\objects\items\candle.dm:41:error: : invalid expression` on the line that only has `if(lit)`, which does work normally if the spawn(0) is there

Comment (Eneocho) [reply code/game/objects/items/candle.dm]: wait nevermind, it works after fixing some weird indentation issues on my side
removing

**Comment by Kurfursten** [comment]
I like your idea to have varying levels of flicker power, which means that multiple candles will flicker slightly differently

**Pull request merged**